utilities/src/main/java/org/craftercms/commons/aws/AwsUtils.java

```diff
@@ -36,95 +37,56 @@ public final class AwsUtils {
-     * @param client          The {@link AmazonS3} client
-     * @param executorFactory The {@link ExecutorFactory} to use to create a {@link TransferManager}
-     * @param sourceBucket    The source bucket
-     * @param sourceBaseKey   The base key in the source bucket (i.e. prefix for all paths)
-     * @param targetBucket    The target bucket
-     * @param targetBaseKey   The base key in the target bucket (i.e. prefix for all paths)
-     * @param paths           The list of paths to copy
-     * @throws InterruptedException If the operation is interrupted while copying the objects
+     * @param client                The {@link S3AsyncClient} client
+     * @param threadPoolExecutor    The thread pool executor
+     * @param sourceBucket          The source bucket
+     * @param sourceBaseKey         The base key in the source bucket (i.e. prefix for all paths)
+     * @param targetBucket          The target bucket
+     * @param targetBaseKey         The base key in the target bucket (i.e. prefix for all paths)
+     * @param paths                 The list of paths to copy
      */
-    public static void copyObjects(AmazonS3 client, ExecutorFactory executorFactory, String sourceBucket, String sourceBaseKey,
-                                   String targetBucket, String targetBaseKey, List<String> paths) throws InterruptedException {
+    public static void copyObjects(S3AsyncClient client, ThreadPoolExecutor threadPoolExecutor, String sourceBucket, String sourceBaseKey,
+                                   String targetBucket, String targetBaseKey, List<String> paths) {
 
-        TransferManager transferManager = buildTransferManager(client, executorFactory);
+        S3TransferManager transferManager = buildTransferManager(client);
         try {
-            CountDownLatch doneSignal = new CountDownLatch(paths.size());
             logger.debug("Copying {} objects from '{}' to '{}'", paths.size(), sourceBucket, targetBucket);
+            List<CompletableFuture<Void>> futures = new ArrayList<>();
             for (String path : paths) {
-                logger.debug("Copying '{}' from '{}' to '{}'", path, sourceBucket, targetBucket);
-                String sourceKey = sourceBaseKey + path;
-                String targetKey = targetBaseKey + path;
-                CopyObjectRequest copyRequest = new CopyObjectRequest()
-                        .withSourceBucketName(sourceBucket)
-                        .withSourceKey(sourceKey)
-                        .withDestinationBucketName(targetBucket)
-                        .withDestinationKey(targetKey);
-                transferManager.copy(copyRequest, new MultiOperationTransferStateChangeListener(sourceBucket, sourceKey, targetBucket, targetKey, doneSignal));
+                futures.add(CompletableFuture.runAsync(() -> {
+                    logger.debug("Copying '{}' from '{}' to '{}'", path, sourceBucket, targetBucket);
+                    String sourceKey = sourceBaseKey + path;
```

Comment:
Might be good to add some defensive coding and make sure that `path` has a `/` at the start, and if not, we add it. I think we had a commons function that does a join adding a separator character when missing

Reply:
I added a util `s3KeyFromPath` with a test case for this.